source/agora/node/Validator.d

```diff
@@ -51,24 +58,75 @@ public class Validator : FullNode, API
     /// Nominator instance
     protected Nominator nominator;
 
+    /// The current quorum configuration (changes as active enrollments change)
+    protected QuorumConfig qc;
+
+    /// The current required set of peer keeys to connect to
+    protected Set!PublicKey required_peer_keys;
+
     /// Ctor
     public this (const Config config)
     {
         assert(config.node.is_validator);
-        super(config);
+        super(config, &this.onValidatorsChanged);
+
+        this.nominator = this.getNominator(this.network,
+            this.config.node.key_pair, this.ledger, this.taskman);
+        this.onValidatorsChanged();
+    }
+
+    /***************************************************************************
+
+        Called when the active validator set has changed after a block
+        was externalized.
 
-        // instantiating Nominator can fail if the quorum configuration
-        // fails the checkSanity() test, and we must release resources.
-        scope (failure)
+        Regenerates the quorum set config and updates the Nominator
+        with the new quorum set.
+
+        The background network discovery task will automatically attempt to
+        find the Validator nodes in the quorum set configuration and connect
+        to them.
+
+    ***************************************************************************/
+
+    private void onValidatorsChanged () nothrow @trusted
+    {
+        try
+        {
+            this.qc = this.buildQuorumConfig();
+            this.nominator.setQuorumConfig(this.qc);
+            buildRequiredKeys(this.config.node.key_pair.address, this.qc,
+                this.required_peer_keys);
+        }
+        catch (Exception ex)
         {
-            this.pool.shutdown();
-            this.utxo_set.shutdown();
-            this.enroll_man.shutdown();
+            log.fatal("onValidatorsChanged() failed: {}", ex);
+            abort();
         }
+    }
 
-        this.nominator = this.getNominator(this.network,
-            this.config.node.key_pair, this.ledger, this.taskman,
-            this.config.quorum);
+    /***************************************************************************
+
+        Generate the quorum configuration for this node based on the
+        blockchain state (enrollments).
+
+        Returns:
+            the generated quorum configuration
+
+    ***************************************************************************/
+
+    private QuorumConfig buildQuorumConfig ()
+    {
+        Enrollment[] enrollments;
+        if (!this.enroll_man.getValidators(enrollments) ||
+            enrollments.length == 0)
+        {
+            log.fatal("Could not retrieve enrollments / no enrollments found");
+            assert(0);
+        }
+
+        return .buildQuorumConfig(this.config.node.key_pair.address,
+            enrollments, this.utxo_set.getUTXOFinder());
```

Comment:
agora.consensus.Quorum.buildQuorumConfig ( const ref PublicKey node_key,
    in Enrollment[] enrolls, in UTXOFinder finder )
I think it's pointing here.
There's no problem. But The code is difficult to read.